src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java

```diff
@@ -0,0 +1,328 @@
+
+    this.currentConnection.setConnectionLifecycleInterceptor(
+        new ConnectionProxyLifecycleInterceptor(this.pluginManager));
+  }
+
+  /**
+   * Checks if connection is associated with Aurora cluster and instantiates a new
+   * AuroraConnectionProxy if needed. Otherwise it returns a single-host connection.
+   *
+   * @param connectionUrl {@link ConnectionUrl} instance containing the lists of hosts available to
+   *     switch on.
+   * @throws SQLException if an error occurs
+   */
+  public static JdbcConnection autodetectClusterAndCreateProxyInstance(ConnectionUrl connectionUrl)
+      throws SQLException {
+
+    boolean pluginsEnabled = Boolean.parseBoolean(
+      connectionUrl.getConnectionArgumentsAsProperties()
+        .getProperty(PropertyKey.useConnectionPlugins.getKeyName(), Boolean.toString(true)));
+
+    if (pluginsEnabled) {
+      final ConnectionProxy connProxy = new ConnectionProxy(connectionUrl);
+      return (JdbcConnection)
+          java.lang.reflect.Proxy.newProxyInstance(
+              JdbcConnection.class.getClassLoader(),
+              new Class<?>[] {JdbcConnection.class},
+              connProxy);
+    }
+
+    final IConnectionProvider connectionProvider = new BasicConnectionProvider();
+    return connectionProvider.connect(connectionUrl.getMainHost());
+  }
+
+  /**
+   * Instantiates a new {@link ConnectionProxy}.
+   *
+   * @param connectionUrl {@link ConnectionUrl} instance containing the lists of hosts available to
+   *     switch on.
+   * @throws SQLException if an error occurs
+   */
+  public static JdbcConnection createProxyInstance(ConnectionUrl connectionUrl)
+      throws SQLException {
+    IConnectionProvider connectionProvider = new BasicConnectionProvider();
+    final ConnectionProxy connProxy = new ConnectionProxy(connectionUrl, connectionProvider.connect(connectionUrl.getMainHost()));
+
+    return (JdbcConnection)
+        java.lang.reflect.Proxy.newProxyInstance(
+            JdbcConnection.class.getClassLoader(),
+            new Class<?>[] {JdbcConnection.class},
+            connProxy);
+  }
+
+  @Override
+  public JdbcConnection getCurrentConnection() {
+    return this.currentConnection;
+  }
+
+  @Override
+  public HostInfo getCurrentHostInfo() {
+    return this.currentHostInfo;
+  }
+
+  @Override
+  public void setCurrentConnection(JdbcConnection connection, HostInfo info) {
+    try {
+      if (this.currentConnection != null && !this.currentConnection.isClosed()) {
+        this.currentConnection.close();
+      }
+    } catch (SQLException sqlEx) {
+      // ignore
+    }
+
+    this.currentConnection = connection;
+    this.currentHostInfo = info;
+  }
+
+  @Override
+  public synchronized Object invoke(Object proxy, Method method, Object[] args)
+      throws Throwable {
+    final String methodName = method.getName();
+
+    if (isDirectExecute(methodName, args)) {
+      return executeMethodDirectly(methodName, args);
+    }
```

Comment:
Hey @ColinKYuen I am confused about this section of the code. Can you help me understand it? There are ClientPreparedStatements created with this ConnectionProxy object. If we execute hashCode() and equals() directly on the proxy object, it would mean that multiple statments created by the same ConnectionProxy would collid in java hash collections. Is this expected? Hibernate is registering my ClientPreparedStatements but it is failing due to finding that a newly registered statement already exists in its internal set even though the already-registered statement is not the same. It only appears to be the same since equals and hashCode just use the same ConnectionProxy. I think this is a bug.